### ISSUE TITLE
loss and reward

### ISSUE THREAD

**Issue by mattinjersey** [Unity-Technologies/ml-agents#1677]
while it is training, the system only outputs the reward in the command line window. Can it be set to also display the loss? Is the loss important for training or is it only training based on the reward?

**Comment by MarcoMeter** [comment]
Use tensorboard to monitor the loss.
https://github.com/Unity-Technologies/ml-agents/blob/master/docs/Using-Tensorboard.md

**Comment by ervteng** [comment]
@MarcoMeter is right - use Tensorboard to monitor the loss. 
The loss is useful in debugging training - it is what we are trying to minimize during training, and is determined by the rewards. It is very generally the difference between "what we want the agent to know" and "what it actually knows."

Some more guidance on the various losses here: https://github.com/Unity-Technologies/ml-agents/docs/Getting-Started-with-Balance-Ball.md

**Comment by mattinjersey** [comment]
agreed...then I am wondering if we could see it in the command window instead of having to use the tensorboard. For example typically when I use PyTorch or Tensorflow for training, I set it to display the loss at each epoch. Right now it only displays the reward and then you have to go to Tensorboard to get the loss.

**Comment by ervteng** [comment]
We don't do it as it's a less intuitive metric (especially for users without an ML background). But it's pretty easy to add it to trainer.py. Should be a 2 line change. In the `write_summary` function you can add loss as a metric to be printed to the screen. 

https://github.com/Unity-Technologies/ml-agents/blob/master/ml-agents/mlagents/trainers/trainer.py

**Comment by mattinjersey** [comment]
ok. I don't see a variable "loss" in that python program?

**Comment by MarcoMeter** [comment]
There are two losses, one for the policy and one for the value estimates.

**Comment by mattinjersey** [comment]
ok could you advise me on the two lines of code referred to above...what actually are the two lines. Thx.

**Comment by ervteng** [comment]
Something like adding `mean_loss = np.mean(self.stats['Losses/Policy Loss'])` after line 163, and adding mean loss into the logger.info call. I haven't tested this but I don't see why it wouldn't work. I'd still recommend using Tensorboard though, as that requires no additional work.

**Comment by mattinjersey** [comment]
can you provide a more clear definition of the value loss and the policy loss

**Comment by mattinjersey** [comment]
another question...could we have the python publish the "advantage" or the output from the critic so that I could see it in real-time from the unity? Does the monitor do this?

**Comment by ervteng** [comment]
Hi Matt, we don't currently support reporting stats from Python back to Unity. You might be able to implement something using the custom Protobuf PR (https://github.com/Unity-Technologies/ml-agents/pull/1595).

**Comment by ervteng** [comment]
Closing this issue due to inactivity. Feel free to open a new issue if you're still having problems. Thanks!
